File: backend/src/main/java/com/dangeremergence/model/CommunityPost.java
package com.dangeremergence.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.UUID;

/**
 * Represents a community post with photo/video media.
 * Users can post incident media from their environment for
 * crowdsourced situational awareness.
 */
@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
@Entity
@Table(name = "community_posts")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class CommunityPost {

    public enum MediaType {
        image, video
    }

    public enum PostStatus {
        active, flagged, removed
    }

    @Id
    @Column(length = 36)
    private String id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(columnDefinition = "TEXT")
    private String caption;

    @Column(name = "media_url", nullable = false, length = 512)
    private String mediaUrl;

    @Column(name = "media_type", nullable = false, length = 20)
    @Enumerated(EnumType.STRING)
    private MediaType mediaType;

    private Double latitude;

    private Double longitude;

    @Column(name = "location_name", length = 255)
    private String locationName;

    @Column(name = "is_anonymous")
    private boolean isAnonymous;

    @Column(nullable = false, length = 20)
    @Enumerated(EnumType.STRING)
    private PostStatus status;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @PrePersist
    public void ensureId() {
        if (id == null) {
            id = UUID.randomUUID().toString();
        }
        if (status == null) {
            status = PostStatus.active;
        }
        if (mediaType == null) {
            mediaType = MediaType.image;
        }
    }
}

File: frontend/lib/modules/community/models/community_post.dart
/// Model representing a community post from the feed.
class CommunityPost {
  final String id;
  final String? caption;
  final String mediaUrl;
  final String mediaType; // 'image' or 'video'
  final double? latitude;
  final double? longitude;
  final String? locationName;
  final bool isAnonymous;
  final String createdAt;
  final int likeCount;
  final int commentCount;
  final int shareCount;
  final bool likedByMe;
  final bool favoritedByMe;
  final Map<String, dynamic> user;

  CommunityPost({
    required this.id,
    this.caption,
    required this.mediaUrl,
    required this.mediaType,
    this.latitude,
    this.longitude,
    this.locationName,
    this.isAnonymous = false,
    required this.createdAt,
    this.likeCount = 0,
    this.commentCount = 0,
    this.shareCount = 0,
    this.likedByMe = false,
    this.favoritedByMe = false,
    required this.user,
  });

  factory CommunityPost.fromJson(Map<String, dynamic> json) {
    return CommunityPost(
      id: json['id'] as String,
      caption: json['caption'] as String?,
      mediaUrl: json['mediaUrl'] as String,
      mediaType: json['mediaType'] as String,
      latitude: (json['latitude'] as num?)?.toDouble(),
      longitude: (json['longitude'] as num?)?.toDouble(),
      locationName: json['locationName'] as String?,
      isAnonymous: json['isAnonymous'] as bool? ?? false,
      createdAt: json['createdAt'] as String,
      likeCount: json['likeCount'] as int? ?? 0,
      commentCount: json['commentCount'] as int? ?? 0,
      shareCount: json['shareCount'] as int? ?? 0,
      likedByMe: json['likedByMe'] as bool? ?? false,
      favoritedByMe: json['favoritedByMe'] as bool? ?? false,
      user: json['user'] as Map<String, dynamic>,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'caption': caption,
      'mediaUrl': mediaUrl,
      'mediaType': mediaType,
      'latitude': latitude,
      'longitude': longitude,
      'locationName': locationName,
      'isAnonymous': isAnonymous,
      'createdAt': createdAt,
      'likeCount': likeCount,
      'commentCount': commentCount,
      'shareCount': shareCount,
      'likedByMe': likedByMe,
      'favoritedByMe': favoritedByMe,
      'user': user,
    };
  }
}
